File: tests/integration/src/fensu/rules/exemplars/test_reachability.py
"""Real native/exemplar parity over production graph inputs."""

from pathlib import Path

import pytest

from fensu.config.main.build_config import build_config
from fensu.config.models import Config
from fensu.discovery.main.discover_files import discover_files
from fensu.discovery.models import DiscoveredTree
from fensu.evaluation.main.evaluate import evaluate
from fensu.evaluation.models import EvaluationResult
from fensu.rules.authoring.main._resolve_rule_spec import resolve_rule_spec
from fensu.rules.authoring.models import RuleSpec
from fensu.rules.exemplars.constants import NATIVE_CUSTOM_RULE_EQUIVALENTS
from fensu.rules.layers.main._dead_code_rules import dead_code_rules
from tests.integration.src.fensu.rules.exemplars._test_types import ReachabilityTestCase


@pytest.mark.parametrize(
    "test_case",
    [
        ReachabilityTestCase(
            "module-level alias preserves member references",
            (
                (
                    "src/orders/__main__.py",
                    "import orders.handlers as handlers\nalias = handlers\nalias.run_orders()\n",
                ),
                ("src/orders/handlers.py", "def run_orders(): pass\ndef unused(): pass\n"),
            ),
            (1, 0, 0),
        ),
        ReachabilityTestCase(
            "empty production surface still checks roots",
            (),
            (0, 1, 0),
            roots=(("orders.missing", "*"),),
        ),
        ReachabilityTestCase(
            "dead helper chain",
            (("src/orders/internal.py", "def tail(): pass\ndef head(): tail()\n"),),
            (2, 0, 1),
        ),
        ReachabilityTestCase(
            "import executes without rooting definitions",
            (
                ("src/orders/__init__.py", "from . import internal\n"),
                ("src/orders/internal.py", "def orphan(): pass\n"),
            ),
            (1, 0, 0),
        ),
        ReachabilityTestCase(
            "parameter shadows imported callable",
            (
                (
                    "src/orders/cli.py",
                    "from .helpers import live as call\nfrom .helpers import unused\ndef main(unused):\n    call()\n    unused()\n",
                ),
                ("src/orders/helpers.py", "def live(): pass\ndef unused(): pass\n"),
            ),
            (1, 0, 0),
            "[project.scripts]\norders = 'orders.cli:main'\n",
        ),
        ReachabilityTestCase(
            "constant identity and deferred initializer",
            (
                ("src/orders/__init__.py", "from .a import VALUE\n__all__ = ['VALUE']\n"),
                ("src/orders/a.py", "VALUE = 1\n"),
                ("src/orders/b.py", "VALUE = 2\nDEAD = lambda: DEAD\n"),
            ),
            (2, 0, 1),
        ),
        ReachabilityTestCase(
            "recursive annotation is not a root",
            (
                ("src/orders/__init__.py", "from . import shapes\n"),
                (
                    "src/orders/shapes.py",
                    "class Node:\n    child: 'Node'\ndef recursive() -> 'recursive': pass\n",
                ),
            ),
            (2, 0, 0),
        ),
        ReachabilityTestCase(
            "literal importlib direct member",
            (
                (
                    "src/orders/__main__.py",
                    "import importlib\nimportlib.import_module('orders.handlers').run_orders()\n",
                ),
                ("src/orders/handlers.py", "def run_orders(): pass\ndef unused(): pass\n"),
            ),
            (1, 0, 0),
        ),
        ReachabilityTestCase(
            "arbitrary decorator is not registration",
            (
                ("src/orders/__init__.py", "from . import internal\n"),
                (
                    "src/orders/internal.py",
                    "from functools import lru_cache\n@lru_cache\ndef unused(): pass\n",
                ),
            ),
            (1, 0, 0),
        ),
        ReachabilityTestCase(
            "public and private main placement",
            (
                ("src/orders/main/_parse.py", "def parse(): pass\n"),
                ("src/orders/main/parse.py", "def parse(): pass\n"),
            ),
            (2, 0, 2),
        ),
        ReachabilityTestCase(
            "initializers execute but lambdas defer",
            (
                ("src/orders/__init__.py", "from . import internal\n"),
                (
                    "src/orders/internal.py",
                    "def setup(): pass\ndef deferred(): pass\nUNUSED = setup()\nLATER = lambda: deferred()\n",
                ),
            ),
            (3, 0, 0),
        ),
        ReachabilityTestCase(
            "atexit decorator registers callback",
            (
                ("src/orders/__init__.py", "from . import callbacks\n"),
                (
                    "src/orders/callbacks.py",
                    "from atexit import register as on_exit\ndef helper(): pass\n@on_exit\ndef shutdown(): helper()\n",
                ),
            ),
            (0, 0, 0),
        ),
        ReachabilityTestCase(
            "nested initializer exports",
            (
                (
                    "src/orders/api/__init__.py",
                    "from .parser import parse_orders\n__all__ = ['parse_orders']\n",
                ),
                ("src/orders/api/parser.py", "def parse_orders(): pass\n"),
            ),
            (0, 0, 0),
        ),
        ReachabilityTestCase(
            "literal importlib local alias",
            (
                (
                    "src/orders/cli.py",
                    "from importlib import import_module as load\ndef main():\n    handlers = load('orders.handlers')\n    handlers.run_orders()\n",
                ),
                ("src/orders/handlers.py", "def run_orders(): pass\ndef unused(): pass\n"),
            ),
            (1, 0, 0),
            "[project.scripts]\norders = 'orders.cli:main'\n",
        ),
        ReachabilityTestCase(
            "application registration depends on application",
            (
                (
                    "src/orders/web.py",
                    "from fastapi import FastAPI\napp = FastAPI()\ndef helper(): pass\n@app.get('/orders')\ndef list_orders(): helper()\ndef unused(): pass\n",
                ),
            ),
            (1, 0, 0),
            "[project.entry-points.'orders.apps']\napp = 'orders.web:app'\n",
        ),
        ReachabilityTestCase(
            "uncalled closure does not retain helper",
            (
                (
                    "src/orders/cli.py",
                    "def helper(): pass\ndef main():\n    def never_called(): helper()\n    return 1\n",
                ),
            ),
            (1, 0, 0),
            "[project.scripts]\norders = 'orders.cli:main'\n",
        ),
        ReachabilityTestCase(
            "returned closure retains helper",
            (
                (
                    "src/orders/cli.py",
                    "def helper(): pass\ndef main():\n    def callback(): helper()\n    return callback\n",
                ),
            ),
            (0, 0, 0),
            "[project.scripts]\norders = 'orders.cli:main'\n",
        ),
        ReachabilityTestCase(
            "cross-module application callbacks",
            (
                (
                    "src/orders/api/callbacks.py",
                    "from .application import app\n@app.get('/orders')\ndef list_orders(): pass\n",
                ),
                (
                    "src/orders/api/application.py",
                    "from fastapi import FastAPI\napp = FastAPI()\nfrom . import callbacks\n",
                ),
            ),
            (0, 0, 0),
            "[project.entry-points.'orders.apps']\napp = 'orders.api.application:app'\n",
        ),
        ReachabilityTestCase(
            "conditional aliases preserve both bindings",
            (
                (
                    "src/orders/cli.py",
                    "if condition:\n    from .a import parse as handler\nelse:\n    from .b import parse as handler\ndef main(): handler()\n",
                ),
                ("src/orders/a.py", "def parse(): pass\n"),
                ("src/orders/b.py", "def parse(): pass\n"),
            ),
            (0, 0, 0),
            "[project.scripts]\norders = 'orders.cli:main'\n",
        ),
        ReachabilityTestCase(
            "final import shadows declaration",
            (
                (
                    "src/orders/cli.py",
                    "def handler(): pass\nfrom .handlers import run_orders as handler\ndef main(): handler()\n",
                ),
                ("src/orders/handlers.py", "def run_orders(): pass\n"),
            ),
            (1, 0, 0),
            "[project.scripts]\norders = 'orders.cli:main'\n",
        ),
        ReachabilityTestCase(
            "wildcard import resolves actual references",
            (
                ("src/orders/cli.py", "from .handlers import *\ndef main(): run_orders()\n"),
                ("src/orders/handlers.py", "def run_orders(): pass\ndef unused(): pass\n"),
            ),
            (1, 0, 0),
            "[project.scripts]\norders = 'orders.cli:main'\n",
        ),
        ReachabilityTestCase(
            "entry and child module have separate identities",
            (
                ("src/orders/api/__init__.py", "def cli(): pass\n"),
                ("src/orders/api/cli.py", "def main(): pass\n"),
            ),
            (1, 0, 1),
            "[project.scripts]\norders = 'orders.api:cli'\n",
        ),
        ReachabilityTestCase(
            "facade wrapper reaches same-named child",
            (
                (
                    "src/orders/api/__init__.py",
                    "def cli():\n    from .cli import main\n    main()\n",
                ),
                ("src/orders/api/cli.py", "def main(): pass\ndef unused(): pass\n"),
            ),
            (1, 0, 0),
            "[project.scripts]\norders = 'orders.api:cli'\n",
        ),
        ReachabilityTestCase(
            "global read before rebinding",
            (
                (
                    "src/orders/cli.py",
                    "from .handlers import run_orders\ndef main():\n    global run_orders\n    run_orders()\n    run_orders = None\n",
                ),
                ("src/orders/handlers.py", "def run_orders(): pass\n"),
            ),
            (0, 0, 0),
            "[project.scripts]\norders = 'orders.cli:main'\n",
        ),
        ReachabilityTestCase(
            "script executes parents and quoted annotations",
            (
                ("src/orders/service/__init__.py", "def setup(): pass\nsetup()\n"),
                (
                    "src/orders/service/cli.py",
                    "class Result: pass\ndef main() -> 'list[Result]': return []\n",
                ),
            ),
            (0, 0, 0),
            "[project.scripts]\norders = 'orders.service.cli:main'\n",
        ),
        ReachabilityTestCase(
            "unrooted cycle",
            (
                (
                    "src/orders/internal.py",
                    "def first():\n    second()\ndef second():\n    first()\n",
                ),
            ),
            (2, 0, 1),
        ),
        ReachabilityTestCase(
            "script aliases and unused declaration",
            (
                (
                    "src/orders/cli.py",
                    "from orders.helpers import live as alias\ndef main():\n    alias()\n",
                ),
                ("src/orders/helpers.py", "def live():\n    pass\ndef dead():\n    pass\n"),
            ),
            (1, 0, 0),
            "[project.scripts]\norders = 'orders.cli:main'\n",
        ),
        ReachabilityTestCase(
            "reasoned root and stale root",
            (("src/orders/internal.py", "def entry():\n    helper()\ndef helper():\n    pass\n"),),
            (0, 1, 0),
            roots=(("orders.internal", "entry"), ("orders.missing", "*")),
        ),
        ReachabilityTestCase(
            "tests do not root main",
            (
                ("src/orders/main/parse.py", "def parse():\n    pass\n"),
                ("tests/test_parse.py", "from orders.main.parse import parse\nparse()\n"),
            ),
            (1, 0, 1),
        ),
        ReachabilityTestCase(
            "public reexport",
            (
                ("src/orders/__init__.py", "from orders.internal import api\n__all__ = ['api']\n"),
                ("src/orders/internal.py", "def api():\n    pass\ndef dead():\n    pass\n"),
            ),
            (1, 0, 0),
        ),
        ReachabilityTestCase(
            "plugin annotations defaults and contracts",
            (
                (
                    "src/orders/plugin.py",
                    "class Model:\n    def contract(self):\n        helper()\ndef helper():\n    pass\nDEFAULT = Model()\ndef entry(value: 'Model' = DEFAULT):\n    pass\n",
                ),
            ),
            (0, 0, 0),
            "[project.entry-points.'orders.plugins']\nplugin = 'orders.plugin:entry'\n",
        ),
        ReachabilityTestCase(
            "main literal dynamic import",
            (
                (
                    "src/orders/__main__.py",
                    "import importlib\nplugin = importlib.import_module('orders.plugin')\nplugin.entry()\n",
                ),
                ("src/orders/plugin.py", "def entry():\n    pass\ndef dead():\n    pass\n"),
            ),
            (1, 0, 0),
        ),
    ],
    ids=lambda case: case.description,
)
def test_given_production_graph_when_evaluating_then_public_traversal_matches_native(
    test_case: ReachabilityTestCase,
    tmp_path: Path,
) -> None:
    (tmp_path / "src/orders").mkdir(parents=True, exist_ok=True)
    for relative, source in test_case.files:
        path: Path = tmp_path / relative
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(source)
    (tmp_path / "pyproject.toml").write_text(test_case.metadata)
    config: Config = build_config(
        raw={
            "roots": ["src/orders"],
            "tests": ["tests"],
            "dead_code": {
                "enabled": True,
                "roots": [
                    {"modules": [module], "symbols": [symbol], "reason": "External registration."}
                    for module, symbol in test_case.roots
                ],
            },
        }
    )
    tree: DiscoveredTree = discover_files(config=config, repo_root=tmp_path)
    counts: list[int] = []
    for core in dead_code_rules():
        exemplar: RuleSpec = resolve_rule_spec(value=NATIVE_CUSTOM_RULE_EQUIVALENTS[core.code])
        native: EvaluationResult = evaluate(tree=tree, ruleset=(core,), config=config)
        custom: EvaluationResult = evaluate(tree=tree, ruleset=(exemplar,), config=config)
        native_findings: list[tuple[str, int | None, int | None, str]] = sorted(
            (str(f.path), f.line, f.column, f.message) for f in native.faults
        )
        custom_findings: list[tuple[str, int | None, int | None, str]] = sorted(
            (str(f.path), f.line, f.column, f.message) for f in custom.faults
        )
        assert custom_findings == native_findings
        counts.append(len(native.faults))
    assert tuple(counts) == test_case.expected_counts


if __name__ == "__main__":
    pytest.main([__file__, "-vv"])

File: crates/fensu-facts/tests/dead_code/test_reachability.rs
use fensu_facts::dead_code::main::analyze::analyze;
use fensu_facts::dead_code::models::{ProjectEntryPoint, Root};

use crate::helpers::{module, package};
use crate::test_types::ReachabilityTestCase;

#[test]
fn given_python_entry_mechanisms_when_traversing_then_only_production_reachability_keeps_definitions(
) {
    let test_cases = [
        ReachabilityTestCase {
            description: "tuple and list assignments execute their RHS",
            modules: vec![module("orders", "from . import assignments\n"), module("orders.assignments", "def tuple_setup(): return (1, 2)\ndef list_setup(): return (3, 4)\nfirst, second = tuple_setup()\n[third, fourth] = list_setup()\n")],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "attribute and subscript assignments execute RHS and target expressions",
            modules: vec![module("orders", "from . import assignments\n"), module("orders.assignments", "def attribute_setup(): return 1\ndef subscript_setup(): return 1\ndef store(): return object()\ndef items(): return {}\ndef index(): return 0\nstore().value = attribute_setup()\nitems()[index()] = subscript_setup()\n")],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "annotated assignments execute targets and RHS",
            modules: vec![module("orders", "from . import assignments\n"), module("orders.assignments", "def attribute_setup(): return 1\ndef subscript_setup(): return 1\ndef store(): return object()\ndef items(): return {}\ndef index(): return 0\nstore().value: int = attribute_setup()\nitems()[index()]: int = subscript_setup()\n")],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "augmented assignments read targets and execute RHS",
            modules: vec![module("orders", "from . import assignments\n"), module("orders.assignments", "def attribute_setup(): return 1\ndef subscript_setup(): return 1\ndef count_setup(): return 1\ndef store(): return object()\ndef items(): return {}\ndef index(): return 0\nstore().value += attribute_setup()\nitems()[index()] += subscript_setup()\ncount = 0\ncount += count_setup()\n")],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "methods skip class bindings while headers and class bodies retain them",
            modules: vec![module("orders", "from .classes import Public\n"), module("orders.classes", "def helper(): pass\ndef header(): pass\ndef decorate(): pass\nclass Public:\n    helper = None\n    header = lambda: None\n    decorate = lambda method: method\n    header()\n    @decorate\n    def run(self, value=header()): return helper()\n")],
            expected_dead: &["orders.classes:header", "orders.classes:decorate"],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "methods preserve enclosing function bindings and their own local shadows",
            modules: vec![module("orders", "from .classes import factory\n"), module("orders.classes", "def helper(): pass\ndef local_helper(): pass\ndef factory():\n    def helper(): return 1\n    class Public:\n        helper = None\n        def run(self): return helper()\n        def local(self):\n            local_helper = lambda: None\n            return local_helper()\n    return Public\n")],
            expected_dead: &["orders.classes:helper", "orders.classes:local_helper"],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "package wildcard exports traverse facades and terminate cycles",
            modules: vec![module("orders", "from .facade import *\n"), module("orders.facade", "from .handlers import *\n"), module("orders.handlers", "from .facade import *\ndef run_orders(): pass\ndef _private(): pass\n")],
            expected_dead: &["orders.handlers:_private"],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "wildcard exports respect explicit all at an intermediate facade",
            modules: vec![module("orders", "from .facade import *\n"), module("orders.facade", "from .handlers import *\n__all__ = ['run_orders']\n"), module("orders.handlers", "def run_orders(): pass\ndef unused(): pass\n")],
            expected_dead: &["orders.handlers:unused"],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "empty all blocks wildcard exports",
            modules: vec![module("orders", "from .facade import *\n"), module("orders.facade", "from .handlers import *\n__all__ = []\n"), module("orders.handlers", "def unused(): pass\n")],
            expected_dead: &["orders.handlers:unused"],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "orphan and unreachable helper chain",
            modules: vec![module("orders", ""), module("orders.internal", "def _orphan(): pass\ndef tail(): pass\ndef head(): tail()\n")],
            expected_dead: &["orders.internal:", "orders.internal:_orphan", "orders.internal:tail", "orders.internal:head"],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "import executes a module without rooting every definition",
            modules: vec![module("orders", "from . import internal\n"), module("orders.internal", "def _orphan(): pass\n")],
            expected_dead: &["orders.internal:_orphan"],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "dead import cycle",
            modules: vec![module("orders", ""), module("orders.a", "from .b import second\ndef first(): second()\n"), module("orders.b", "from .a import first\ndef second(): first()\n")],
            expected_dead: &["orders.a:", "orders.a:first", "orders.b:", "orders.b:second"],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "script aliases and parameter shadowing",
            modules: vec![module("orders", ""), module("orders.cli", "from .internal import live as call\nfrom .internal import unused\ndef main(unused):\n    call()\n    unused()\n"), module("orders.internal", "def live(): pass\ndef unused(): pass\n")],
            entries: &["orders.cli:main"],
            expected_dead: &["orders.internal:unused"],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "public class exports preserve dunders and contracts",
            modules: vec![module("orders", "from .internal import Adapter\n__all__ = ['Adapter']\n"), module("orders.internal", "def helper(): pass\nclass Adapter:\n    def __call__(self): helper()\n    def contract(self): helper()\n")],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "constant identity and dead self initializer",
            modules: vec![module("orders", "from .a import VALUE\n__all__ = ['VALUE']\n"), module("orders.a", "VALUE = 1\n"), module("orders.b", "VALUE = 2\nDEAD = lambda: DEAD\n")],
            expected_dead: &["orders.b:", "orders.b:VALUE", "orders.b:DEAD"],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "configured roots match before reachability",
            modules: vec![module("orders", ""), module("orders.handlers", "def run_orders(): pass\n")],
            roots: vec![Root { modules: vec!["orders.handlers".to_owned()], symbols: vec!["run_*".to_owned()] }],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "recursive annotations do not self-root dead declarations",
            modules: vec![module("orders", "from . import shapes\n"), module("orders.shapes", "class Node:\n    child: 'Node'\ndef recursive() -> 'recursive': pass\n")],
            expected_dead: &["orders.shapes:Node", "orders.shapes:recursive"],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "removed declaration makes its root stale",
            modules: vec![module("orders", "")],
            roots: vec![Root { modules: vec!["orders.handlers".to_owned()], symbols: vec!["run_*".to_owned()] }],
            expected_stale: &[0],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "literal importlib member access",
            modules: vec![module("orders", ""), module("orders.cli", "import importlib\ndef main():\n    importlib.import_module('orders.handlers').run_orders()\n"), module("orders.handlers", "def run_orders(): pass\ndef unused(): pass\n")],
            entries: &["orders.cli:main"], expected_dead: &["orders.handlers:unused"],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "arbitrary decorator is not registration",
            modules: vec![module("orders", "from . import internal\n"), module("orders.internal", "from functools import lru_cache\n@lru_cache\ndef unused(): pass\n")],
            expected_dead: &["orders.internal:unused"],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "private and public main placement are not roots",
            modules: vec![module("orders", ""), module("orders.service.main._parse", "def parse_orders(): pass\n"), module("orders.service.main.parse", "def parse_orders(): pass\n")],
            expected_dead: &["orders.service.main._parse:", "orders.service.main._parse:parse_orders", "orders.service.main.parse:", "orders.service.main.parse:parse_orders"],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "exported and registered main entries",
            modules: vec![module("orders", "from .service.main._parse import parse_orders\n__all__ = ['parse_orders']\n"), module("orders.service.main._parse", "def parse_orders(): pass\n"), module("orders.service.main.parse", "def parse_orders(): pass\n")],
            entries: &["orders.service.main.parse:parse_orders"],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "initializers execute but lambda bodies are deferred",
            modules: vec![module("orders", "from . import internal\n"), module("orders.internal", "def setup(): pass\ndef deferred(): pass\nUNUSED = setup()\nLATER = lambda: deferred()\n")],
            expected_dead: &["orders.internal:deferred", "orders.internal:UNUSED", "orders.internal:LATER"],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "atexit callback registration",
            modules: vec![module("orders", "from . import callbacks\n"), module("orders.callbacks", "from atexit import register as on_exit\ndef helper(): pass\n@on_exit\ndef shutdown(): helper()\n")],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "nested initializer relative imports",
            modules: vec![module("orders", ""), package("orders.api", "from .parser import parse_orders\n__all__ = ['parse_orders']\n"), module("orders.api.parser", "def parse_orders(): pass\n")],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "literal importlib local alias",
            modules: vec![module("orders", ""), module("orders.cli", "from importlib import import_module as load\ndef main():\n    handlers = load('orders.handlers')\n    handlers.run_orders()\n"), module("orders.handlers", "def run_orders(): pass\ndef unused(): pass\n")],
            entries: &["orders.cli:main"], expected_dead: &["orders.handlers:unused"],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "application registrations depend on the live application",
            modules: vec![module("orders", ""), module("orders.web", "from fastapi import FastAPI\napp = FastAPI()\ndef helper(): pass\n@app.get('/orders')\ndef list_orders(): helper()\ndef unused(): pass\n")],
            entries: &["orders.web:app"], expected_dead: &["orders.web:unused"],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "uncalled local closure does not root its helper",
            modules: vec![module("orders", ""), module("orders.cli", "def helper(): pass\ndef main():\n    def never_called(): helper()\n    return 1\n")],
            entries: &["orders.cli:main"], expected_dead: &["orders.cli:helper"],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "cross-module callback registration is independent of discovery order",
            modules: vec![module("orders", ""), module("orders.api.callbacks", "from .application import app\n@app.get('/orders')\ndef list_orders(): pass\n"), module("orders.api.application", "from fastapi import FastAPI\napp = FastAPI()\nfrom . import callbacks\n")],
            entries: &["orders.api.application:app"],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "returned local closure retains its dependencies",
            modules: vec![module("orders", ""), module("orders.cli", "def helper(): pass\ndef main():\n    def callback(): helper()\n    return callback\n")],
            entries: &["orders.cli:main"],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "conditional imports preserve both possible bindings",
            modules: vec![module("orders", ""), module("orders.cli", "if condition:\n    from .a import parse as handler\nelse:\n    from .b import parse as handler\ndef main(): handler()\n"), module("orders.a", "def parse(): pass\n"), module("orders.b", "def parse(): pass\n")],
            entries: &["orders.cli:main"],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "a final import shadows an earlier declaration",
            modules: vec![module("orders", ""), module("orders.cli", "def handler(): pass\nfrom .handlers import run_orders as handler\ndef main(): handler()\n"), module("orders.handlers", "def run_orders(): pass\n")],
            entries: &["orders.cli:main"], expected_dead: &["orders.cli:handler"],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "wildcard imports resolve actual symbol references",
            modules: vec![module("orders", ""), module("orders.cli", "from .handlers import *\ndef main(): run_orders()\n"), module("orders.handlers", "def run_orders(): pass\ndef unused(): pass\n")],
            entries: &["orders.cli:main"], expected_dead: &["orders.handlers:unused"],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "entrypoint declaration is distinct from a same-named child module",
            modules: vec![module("orders", ""), package("orders.api", "def cli(): pass\n"), module("orders.api.cli", "def main(): pass\n")],
            entries: &["orders.api:cli"], expected_dead: &["orders.api.cli:", "orders.api.cli:main"],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "a facade wrapper can reach a same-named child module without merging them",
            modules: vec![module("orders", ""), package("orders.api", "def cli():\n    from .cli import main\n    main()\n"), module("orders.api.cli", "def main(): pass\ndef unused(): pass\n")],
            entries: &["orders.api:cli"], expected_dead: &["orders.api.cli:unused"],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "global reads before rebinding retain the original callable",
            modules: vec![module("orders", ""), module("orders.cli", "from .handlers import run_orders\ndef main():\n    global run_orders\n    run_orders()\n    run_orders = None\n"), module("orders.handlers", "def run_orders(): pass\n")],
            entries: &["orders.cli:main"],
            ..Default::default()
        },
        ReachabilityTestCase {
            description: "script executes parents and preserves quoted type references",
            modules: vec![module("orders", ""), package("orders.service", "def setup(): pass\nsetup()\n"), module("orders.service.cli", "class Result: pass\ndef main() -> 'list[Result]': return []\n")],
            entries: &["orders.service.cli:main"],
            ..Default::default()
        },
    ];
    for test_case in test_cases {
        let entries = test_case
            .entries
            .iter()
            .map(|entry| ProjectEntryPoint {
                kind: "script".to_owned(),
                reference: (*entry).to_owned(),
            })
            .collect::<Vec<_>>();
        let result = analyze(&test_case.modules, &entries, &test_case.roots);
        let dead = result
            .dead
            .into_iter()
            .map(|declaration| {
                format!(
                    "{}:{}",
                    test_case.modules[declaration.module].name, declaration.name
                )
            })
            .collect::<Vec<_>>();
        assert_eq!(dead, test_case.expected_dead, "{}", test_case.description);
        assert_eq!(
            result.stale_roots, test_case.expected_stale,
            "{}",
            test_case.description
        );
    }
}
